Mathlib/Algebra/Category/ModuleCat/TensorHomAdjunction.lean

```diff
@@ -0,0 +1,257 @@
+/-
+Copyright (c) 2023 Jujian Zhang. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Jujian Zhang
+-/
+import Mathlib.LinearAlgebra.TensorProduct
+import Mathlib.Algebra.Category.ModuleCat.Basic
+import Mathlib.CategoryTheory.Adjunction.Limits
+import Mathlib.LinearAlgebra.Finsupp
+
+/-!
+# Tensor-Hom adjunction
+Consider two commutative rings `R` and `S` and `X` an `(R, S)`-bimodule.
+Consider the tensor functor `(X ⊗[R] .)` from the category of left `R`-module to the category of
+right `S`-module and the hom functor `X →ₗ[Sᵐᵒᵖ] .` from the category of right `S`-module to the
+category of left `R`-module. They form an adjunction. In particular we have that
+```
+Hom_Sᵐᵒᵖ(X⊗[R]Y, Z) ≃ Hom_R(Y, Hom_Sᵐᵒᵖ(X, Z))
+```
+
+## Implementation notes
+
+1. Order of arguments
+```
+Hom_Sᵐᵒᵖ(Y⊗[R]X, Z) ≃ Hom_R(Y, Hom_Sᵐᵒᵖ(X, Z))
+```
+is perhaps more natural to work with because all the argument have the same order.
+But currently mathlib4 does not know `Y⊗[R] X` is an right `S`-module as well.
+
+2. Why not use the [`Tower` file](Mathlib/LinearAlgebra/TensorProduct/Tower.lean`)
+
+In our setting `X` is an `(R, S)`-bimodule and `Y` an `R`-module and `Z` an `S`-module
+so to use the `Tower` file, we need `S` to be an `R`-algebra which is a luxury we do not have.
+But note that in `Tower` file, `curry` and `uncurry` are both tri-linear maps. So `Tower` file
+allows interplay of 3 rings which is not allowed in this file.
+
+-/
+
+universe u u' v v' v''
+
+open TensorProduct CategoryTheory
+
+variable (R : Type u) (S : Type u') (X : Type v)
+variable [CommRing R] [Ring S]
+variable [AddCommGroup X] [Module R X] [Module Sᵐᵒᵖ X] [SMulCommClass R Sᵐᵒᵖ X]
+
+local instance hom_bimodule {Y : Type v''} [AddCommGroup Y] [Module Sᵐᵒᵖ Y] :
+    Module R (X →ₗ[Sᵐᵒᵖ] Y) where
```

Comment:
Arguably this should be
```suggestion
    Module Rᵈᵐᵃ (X →ₗ[Sᵐᵒᵖ] Y) where
```
where `ᵈᵐᵃ` is `DomMulAct`. I don't know if `ᵈᵐᵃ` is sufficiently developed for that to work out of the box.

Reply:
By chaning this instance, it would manefest what you just commented [here](https://github.com/leanprover-community/mathlib4/pull/8495#discussion_r1398479089), we have left `R`-module, right `S`-module and `Rᵈᵐᵃ` module (which is just right `R` module?)

Reply:
Yes, because this is really a right-action!

Reply:
Since we are working with functors between left `R`-modules and right `S`-modules, we eventually need it to be `R`-module.

Reply:
Indeed, but I think you can use semilinear maps for that